### PR TITLE
fix os_default_template api_path

### PR DESCRIPTION
**Pull request by mdellweg**

### CODE REVIEW

**Comment by evgeni** [comment]
@SatelliteQE/quality-engineers can I request review from one of you please?

**Comment by mdellweg** [comment]
@renzon will this be merged soon?

**Comment by omaciel** [comment]
@abalakh or @svtkachenko could you guys take a look at this when you get a chance?

File: nailgun/entities.py [hunk 1]
```diff
@@ -4205,27 +4205,61 @@ def download_debug_certificate(self, synchronous=True, **kwargs):
...
 
 
-class OSDefaultTemplate(Entity):
+class OSDefaultTemplate(
+        Entity,
+        EntityCreateMixin,
+        EntityDeleteMixin,
+        EntityReadMixin,
+        EntitySearchMixin,
+        EntityUpdateMixin):
```
Comment (abalakh) [inline nailgun/entities.py]: Please provide some test results: commandline output for each added operation - create/read/search/update/delete to ensure every single operation works and no more workarounds (like custom `update_payload` method you've already implemented) are needed.

File: nailgun/entities.py [hunk 2]
```diff
@@ -4205,27 +4205,61 @@ def download_debug_certificate(self, synchronous=True, **kwargs):
...
+                operatingsystem=self.operatingsystem,
+                # pylint:enable=no-member
+            )
+        if ignore is None:
+            ignore = set()
+        ignore.add('operatingsystem')
+        return super(OSDefaultTemplate, self).read(entity,
+                                                   attrs,
+                                                   ignore,
+                                                   params)
```
Comment (abalakh) [inline nailgun/entities.py]: nitpick: we use different indentation style:
```python
        return super(OSDefaultTemplate, self).read(
            entity, attrs, ignore, params)
```

```python
        self.assertIn(
            'operatingsystems/1/os_default_templates/2',
            entities.OSDefaultTemplate(
                self.cfg, id=2, operatingsystem=1).path()
        )
```
and so on for the rest of PR. It's not a blocker, rather recommendation to keep the code in one style 🙂

File: nailgun/entities.py [hunk 3]
```diff
@@ -4205,27 +4205,61 @@ def download_debug_certificate(self, synchronous=True, **kwargs):
...
+
+    def update_payload(self, fields=None):
+        """wrap payload in 'os_default_template'
+        relates to http://projects.theforeman.org/issues/21169
```
Comment (abalakh) [inline nailgun/entities.py]: nitpick: `s/wrap/Wrap/` , `s/'os_default_template'/``os_default_template``/` and highlight the link:
```
relates to `Redmine #21169`_.

.. _Redmine #21169: http://projects.theforeman.org/issues/21169
```

File: tests/test_entities.py [hunk 4]
```diff
@@ -398,6 +400,20 @@ def test_no_such_path_error(self):
                 with self.assertRaises(NoSuchPathError):
                     entity(self.cfg_610).path(which=which)
 
+    def test_os_default_template(self):
+        """ Test ``nailgun.entities.OSDefaultTemplate.path``
```
Comment (abalakh) [inline tests/test_entities.py]:
```
Test :meth:`nailgun.entities.OSDefaultTemplate.path`
```

File: tests/test_entities.py [hunk 5]
```diff
@@ -1562,6 +1579,23 @@ def test_organization_rh_repo_url(self):
         ).update_payload()
         self.assertIn('redhat_repository_url', payload)
 
+    def test_os_default_template(self):
+        """Check, that os_default_template serves template_kind_id and
+        provisioning_template_id only wrapped in sub dict
+        See: http://projects.theforeman.org/issues/21169
```
Comment (abalakh) [inline tests/test_entities.py]: Same here - `/s/os_default_template/``os_default_template``/`, same for `template_kind_id` and `provisioning_template_id`.
And pls highlight the link 🙂

File: nailgun/entities.py (hunk 1 above)
Comment (mdellweg) [reply nailgun/entities.py]: Where do you want the results? What do you mean by commandline? IPython?
Can you give me some pointers?

Comment (svtkachenko) [reply nailgun/entities.py]: Yes, ipython, `entities.OSDefaultTemplate().create()`, `entities.OSDefaultTemplate(id=id).read()` etc.

Comment (abalakh) [reply nailgun/entities.py]: Just a comment here containing iPython's output should be enough. Pointers - any PR containing new mixins for entities, for example:
https://github.com/SatelliteQE/nailgun/pull/399#issue-222688398
https://github.com/SatelliteQE/nailgun/pull/446#issue-259765994

**Comment by mdellweg** [comment]
```
In [1]: os = entities.OperatingSystem(id=1).read()

In [2]: entities.OSDefaultTemplate(operatingsystem=os).search()
Out[2]: 
[nailgun.entities.OSDefaultTemplate(provisioning_template=nailgun.entities.ProvisioningTemplate(id=40), template_kind=nailgun.entities.TemplateKind(id=5), operatingsystem=nailgun.entities.OperatingSystem(id=1), config_template=nailgun.entities.ConfigTemplate(id=40), id=1),
 nailgun.entities.OSDefaultTemplate(provisioning_template=nailgun.entities.ProvisioningTemplate(id=44), template_kind=nailgun.entities.TemplateKind(id=4), operatingsystem=nailgun.entities.OperatingSystem(id=1), config_template=nailgun.entities.ConfigTemplate(id=44), id=3),
 nailgun.entities.OSDefaultTemplate(provisioning_template=nailgun.entities.ProvisioningTemplate(id=43), template_kind=nailgun.entities.TemplateKind(id=3), operatingsystem=nailgun.entities.OperatingSystem(id=1), config_template=nailgun.entities.ConfigTemplate(id=43), id=4),
 nailgun.entities.OSDefaultTemplate(provisioning_template=nailgun.entities.ProvisioningTemplate(id=42), template_kind=nailgun.entities.TemplateKind(id=1), operatingsystem=nailgun.entities.OperatingSystem(id=1), config_template=nailgun.entities.ConfigTemplate(id=42), id=5),
 nailgun.entities.OSDefaultTemplate(provisioning_template=nailgun.entities.ProvisioningTemplate(id=45), template_kind=nailgun.entities.TemplateKind(id=8), operatingsystem=nailgun.entities.OperatingSystem(id=1), config_template=nailgun.entities.ConfigTemplate(id=45), id=6),
 nailgun.entities.OSDefaultTemplate(provisioning_template=nailgun.entities.ProvisioningTemplate(id=41), template_kind=nailgun.entities.TemplateKind(id=6), operatingsystem=nailgun.entities.OperatingSystem(id=1), config_template=nailgun.entities.ConfigTemplate(id=41), id=8)]

In [3]: odt = entities.OSDefaultTemplate(operatingsystem=os).search()[5]

In [4]: odt.config_template=entities.ProvisioningTemplate(id=97)

In [5]: odt.update()
Out[5]: nailgun.entities.OSDefaultTemplate(provisioning_template=nailgun.entities.ProvisioningTemplate(id=97), template_kind=nailgun.entities.TemplateKind(id=6), operatingsystem=nailgun.entities.OperatingSystem(id=1), config_template=nailgun.entities.ConfigTemplate(id=97), id=8)

In [6]: odt.delete()
Out[6]: 
{u'created_at': u'2017-11-06T17:17:16.829Z',
 u'id': 8,
 u'operatingsystem_id': 1,
 u'provisioning_template_id': 97,
 u'template_kind_id': 6,
 u'updated_at': u'2017-11-06T17:22:35.638Z'}

In [7]: odt.create()
Out[7]: nailgun.entities.OSDefaultTemplate(provisioning_template=nailgun.entities.ProvisioningTemplate(id=97), template_kind=nailgun.entities.TemplateKind(id=6), operatingsystem=nailgun.entities.OperatingSystem(id=1), config_template=nailgun.entities.ConfigTemplate(id=97), id=9)
```

File: nailgun/entities.py (hunk 1 above)
Comment (mdellweg) [reply nailgun/entities.py]: That explains, why i did not find any in the source...
Added at the bottom. 👇

File: tests/test_entities.py [hunk 6]
```diff
@@ -398,6 +400,19 @@ def test_no_such_path_error(self):
                 with self.assertRaises(NoSuchPathError):
                     entity(self.cfg_610).path(which=which)
 
+    def test_os_default_template(self):
+        """ Test :meth:`nailgun.entities.OSDefaultTemplate.path`
```
Comment (abalakh) [inline tests/test_entities.py]: My bad, according to travis log sphinx can't build a link as can't find inherited method. Pls revert back to 
```
``nailgun.entities.OSDefaultTemplate.path``
```

Comment (mdellweg) [reply tests/test_entities.py]: Yea, just figured that out, too. Will revert. Thanks!

**Comment by mdellweg** [comment]
Thanks for your help!

**Comment by omaciel** [comment]
Thank you @mdellweg and @abalakh :)

**Pull request merged**